server/asset/btc/btc_test.go

```diff
@@ -303,73 +303,77 @@ func txOutID(txHash *chainhash.Hash, index uint32) string {
 	return txHash.String() + ":" + strconv.Itoa(int(index))
 }
 
-const optimalFeeRate uint64 = 24
-
-func (*testNode) EstimateSmartFee(confTarget int64, mode *btcjson.EstimateSmartFeeMode) (*btcjson.EstimateSmartFeeResult, error) {
-	optimalRate := float64(optimalFeeRate) * 1e-5
-	// fmt.Println((float64(optimalFeeRate)*1e-5)-0.00024)
-	return &btcjson.EstimateSmartFeeResult{
-		Blocks:  2,
-		FeeRate: &optimalRate,
-	}, nil
-}
-
-// Part of the btcNode interface.
-func (t *testNode) GetTxOut(txHash *chainhash.Hash, index uint32, _ bool) (*btcjson.GetTxOutResult, error) {
-	testChainMtx.RLock()
-	defer testChainMtx.RUnlock()
-	outID := txOutID(txHash, index)
-	out := testChain.txOuts[outID]
-	// Unfound is not an error for GetTxOut.
-	return out, nil
-}
-
-// Part of the btcNode interface.
-func (t *testNode) GetRawTransactionVerbose(txHash *chainhash.Hash) (*btcjson.TxRawResult, error) {
-	testChainMtx.RLock()
-	defer testChainMtx.RUnlock()
-	tx, found := testChain.txRaws[*txHash]
-	if !found {
-		return nil, fmt.Errorf("test transaction not found\n")
-	}
-	return tx, nil
-}
-
-// Part of the btcNode interface.
-func (t *testNode) GetBlockVerbose(blockHash *chainhash.Hash) (*btcjson.GetBlockVerboseResult, error) {
-	testChainMtx.RLock()
-	defer testChainMtx.RUnlock()
-	block, found := testChain.blocks[*blockHash]
-	if !found {
-		return nil, fmt.Errorf("test block not found")
-	}
-	return block, nil
-}
-
-// Part of the btcNode interface.
-func (t *testNode) GetBlockHash(blockHeight int64) (*chainhash.Hash, error) {
-	testChainMtx.RLock()
-	defer testChainMtx.RUnlock()
-	hash, found := testChain.hashes[blockHeight]
-	if !found {
-		return nil, fmt.Errorf("test hash not found")
-	}
-	return hash, nil
-}
-
-// Part of the btcNode interface.
-func (t *testNode) GetBestBlockHash() (*chainhash.Hash, error) {
-	testChainMtx.RLock()
-	defer testChainMtx.RUnlock()
-	bbHash := testBestBlock.hash
-	return &bbHash, nil
-}
-
-func (t *testNode) RawRequest(string, []json.RawMessage) (json.RawMessage, error) {
-	if t.rawErr != nil {
-		return nil, t.rawErr
+func (t *testNode) Shutdown() {}
+
+func (t *testNode) WaitForShutdown() {}
+
+func (t *testNode) RawRequest(_ context.Context, method string, params []json.RawMessage) (json.RawMessage, error) {
+	switch method {
+	case methodGetBestBlockHash:
+		testChainMtx.RLock()
+		defer testChainMtx.RUnlock()
+		bbHash := testBestBlock.hash
+		return json.Marshal(bbHash.String())
+	case methodGetBlock:
+		testChainMtx.RLock()
+		defer testChainMtx.RUnlock()
+		var blkHash string
+		_ = json.Unmarshal(params[0], &blkHash)
+		blockHash, err := chainhash.NewHashFromStr(blkHash)
+		if err != nil {
+			return nil, fmt.Errorf("rawrequest: %v method param unmarshal error: %v",
+				method, err)
+		}
+		block, found := testChain.blocks[*blockHash]
+		if !found {
+			return nil, fmt.Errorf("test block not found")
+		}
+		return json.Marshal(block)
```

Comment:
There are places where we have hacks to modify the `GetBlockVerboseResult` in the `testchain.blocks` map, however a workaround is needed because with this stub, it marshals to json, then the new `(*rpcClient).GetBlockVerbose` uses `call` to unmarshal into a *new* `*GetBlockVerboseResult`, not the same one that's in the cache.

So for example in `TestReorg` > `setSidechainConfs` we have:

```go
			blk, err := btc.node.GetBlockVerbose(hash)
			if err != nil {
				t.Fatalf("error retrieving sidechain block to set confirmations: %v", err)
			}
			// Set Confirmations
			btc.blockCache.mtx.Lock() // read from (*blockCache).add in cache.go
			testChainMtx.Lock()       // field of testChain
			blk.Confirmations = -1
			testChainMtx.Unlock()
			btc.blockCache.mtx.Unlock()
```

Instead, we need to be setting the `Confirmations` field of the blocks in the `testchain` map.  I think this can be done directly (always could have), but let me have a look...

Reply:
Riiiight, good catch, let me try set that directly on the testchain map.